src/pb_memory.cc

```diff
@@ -141,15 +141,36 @@ PbMemory::CopyBuffer(
     kind = cudaMemcpyDeviceToDevice;
   }
 
-  cudaError_t err =
-      cudaMemcpy(dst->DataPtr(), src->DataPtr(), src->ByteSize(), kind);
+  cudaError_t err;
+  if ((src->MemoryType() == TRITONSERVER_MEMORY_GPU &&
+       dst->MemoryType() == TRITONSERVER_MEMORY_GPU) &&
+      (src->MemoryTypeId() != dst->MemoryTypeId())) {
+    err = cudaMemcpyPeer(
+        dst->DataPtr(), dst->MemoryTypeId(), src->DataPtr(),
+        src->MemoryTypeId(), src->ByteSize());
+
+  } else {
+    err = cudaMemcpy(dst->DataPtr(), src->DataPtr(), src->ByteSize(), kind);
+  }
 
   if (err != cudaSuccess) {
     throw PythonBackendException(
         std::string(
             "failed to copy data: " + std::string(cudaGetErrorString(err)))
             .c_str());
   }
+
+  if ((src->MemoryType() == TRITONSERVER_MEMORY_GPU &&
+       dst->MemoryType() == TRITONSERVER_MEMORY_GPU)) {
+    // Synchronize the default stream for d2d copies.
+    // https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-sync
+    err = cudaStreamSynchronize(0);
```

Comment:
Do we need to explicitly use the legacy stream here? CU_STREAM_LEGACY

Reply:
I don't think that it is necessary, since default stream has id 0 and it will behave either like Legacy Default (or PTDS, which we don't use). I think it is better to keep 0. We can get rid of the magic number though, by specifying CUDA_DEFAULT_STREAM for our code.

Reply:
> Do we need to explicitly use the legacy stream here? CU_STREAM_LEGACY

It looks like this maps to 1. I double checked. synchronization here would synchronize with all the blocking streams which is the expected behavior. It doesn't synchronize the non-blocking streams.

> Do you think that this sync here may impact user's code? Should we consider async D2D copy on a separate stream and sync on that stream and not on the default stream?

We can do this in a follow up PR. Once you're PR is merged in we can use those streams instead of default stream.

Reply:
I guess my question comes back to - do we need this to be the legacy stream or is it ok to be a per thread default stream? If I understand correctly - we just need to ensure the copy completes before the host moves on. Is the behavior that cudamemcpy / peermemcpy happens on default stream (either legacy or per thread) - and synchronize(0) will ensure the copy happens before the host moves on?

Reply:
Default stream behaves either like Legacy Default Stream or Per-Thread Default Stream. By default it behaves like Legacy Default. To change this behavior, one needs to compile cuda code with the --default-stream flag. See [stream sync behavior docs](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html#stream-sync-behavior). 


> The behavior that cudamemcpy / peermemcpy happens on default stream
It is true, their async counterparts allow specification of a stream to avoid default stream. 

> and synchronize(0) will ensure the copy happens before the host moves on
This is true, but explicit sync needed only for d2d copy, because according to the docs: https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-sync 
in this case no host-side sync is performed. 
@Tabrizian feel free to correct me

Reply:
Thanks!